torch/_dynamo/metrics_context.py

```diff
@@ -0,0 +1,124 @@
+        a set of metrics, e.g., metrics gathered during a compilation. On exit of the
+        contextmanager, call the provided 'on_exit' function and pass a dictionary of
+        all metrics set during the lifetime of the contextmanager.
+        """
+        self._on_exit = on_exit
+        self._metrics: Dict[str, Any] = {}
+        self._start_time: Dict[str, int] = {}
+        self._recording = False
+
+    def __enter__(self) -> "MetricsContext":
+        """
+        Initialize metrics recording.
+        """
+        if self._recording:
+            raise RuntimeError("Cannot re-enter a MetricsContext")
+        self._metrics = {}
+        self._start_time = {}
+        self._recording = True
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        _traceback: Any,
+    ) -> None:
+        """
+        Post-process metrics where appropriate and call the provided on_exit function.
+        """
+        self._on_exit(self._metrics, exc_type, exc_value)
+        self._recording = False
+
+    def _check(self) -> None:
+        """
+        Raise if we haven't entered the contextmanager.
+        """
+        if not self._recording:
+            raise RuntimeError("Cannot set metrics outside of a MetricsContext")
+
+    def increment(self, metric: str, value: Union[str, float]) -> None:
+        """
+        Increment a metric by a given amount.
+        """
+        self._check()
+        if metric not in self._metrics:
+            self._metrics[metric] = 0
+        self._metrics[metric] += value
+
+    def set(self, metric: str, value: Any) -> None:
+        """
+        Set a metric to a given value. Use set_once to raise if the metric is set
+        more than once.
+        """
+        self._check()
+        self._metrics[metric] = value
+
+    def set_once(self, metric: str, value: Any) -> None:
+        """
+        Set a metric to a given value. Raise if the metrics has been set previously
+        in the current context.
+        """
+        self._check()
+        if metric in self._metrics:
+            raise RuntimeError(
+                f"Metric '{metric}' has already been set in the current context"
+            )
+        self._metrics[metric] = value
+
+    def update(self, values: Dict[str, Any]) -> None:
+        """
+        Set multiple metrics.
+        """
+        self._check()
+        self._metrics.update(values)
+
+    @contextmanager
+    def timed(self, metric: str) -> Generator[Any, None, None]:
+        """
+        Use this context manager to record execution time. Automatically adjusts the
+        recorded time according to the suffix of the provided metric name: "_ns", "_us",
+        "_ms", or "_s" for nanoseconds, microseconds, etc.
+        """
+
+        # To properly record timing when there's recursion, start and stop the timer
```

Comment:
I want to be able to add timing to `PyCodeCache.load_by_key` and that can call itself.